ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +194,82 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
 	return w.stack.PeekAll()
 }
 
-// pollNextBlock polls for the next block header to be added to the block stack.
-// If there are no blocks on the stack, it fetches the first block at the specified
-// `startBlockDepth` supplied at instantiation.
-func (w *Watcher) pollNextBlock() error {
-	var nextBlockNumber *big.Int
-	latestHeader, err := w.stack.Peek()
+// syncChain syncs our local state of the chain to the latest block found via
+// Ethereum RPC
+func (w *Watcher) syncChain() error {
+	latestHeader, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return err
 	}
-	if latestHeader == nil {
-		if w.startBlockDepth == rpc.LatestBlockNumber {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestHeader.Number.Int64() <= lastStoredHeader.Number.Int64() {
+			return nil
+		}
+		numBlocksToFetch = int(latestHeader.Number.Int64() - lastStoredHeader.Number.Int64())
 	}
-	if err != nil {
-		return err
+
+	allEvents := []*Event{}
+	var syncErr error
+	for i := 0; i < numBlocksToFetch; i++ {
+		// Optimization: If numBlocksToFetch is 1, we already know what the nextHeader is, so avoid
+		// fetching it again. If there is more then 1 block to fetch, compute each from the last
+		// stored and fetch it
+		nextHeader := latestHeader
+		if numBlocksToFetch != 1 {
+			lastStoredHeader, syncErr := w.stack.Peek()
```

Comment:
There's a subtle bug here. Because you used the `:=` operator, `syncErr` is redefined as a new variable within the scope of the for loop. Even though the two variables have the same name, changing the value of the one inside of the for loop will not change the one on the outside. See this playground example: https://play.golang.org/p/MsDG77VWgUr.

This is a common gotcha in Go and we should try to avoid declaring and redefining `syncErr` like this altogether if possible. It looks like the main reason you wrote the code this way is to clean up at the end of the for loop by either calling `Reset` or `Checkpoint` and potentially sending events through the `blockFeed`. I wonder if there is another way to structure this code? Maybe we could make use of `defer`?

The presence of this bug also makes me nervous about our test coverage. Can we improve our tests so that we would catch cases like this?

Reply:
Good catch. I've re-written this logic so that we use the regular `err` var when assigning, and only if an error occurs do we assign it to `syncErr` before `break`ing. 

The test coverage for BlockWatcher is at ~80% and almost all lines that aren't covered are error conditions. I'll work on bumping up coverage and will start with a regression test for the above.